velox/expression/tests/ConjunctRewriteTest.cpp

```diff
@@ -0,0 +1,90 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+
+#include "velox/expression/ExprRewriteRegistry.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+namespace facebook::velox::expression {
+namespace {
+
+class ConjunctRewriteTest : public functions::test::FunctionBaseTest {
+ protected:
+  static void SetUpTestCase() {
+    memory::MemoryManager::testingSetInstance(memory::MemoryManager::Options{});
+  }
+
+  void SetUp() override {
+    functions::prestosql::registerAllScalarFunctions("");
+    parse::registerTypeResolver();
+  }
+
+  void TearDown() override {
+    expression::ExprRewriteRegistry::instance().clear();
+  }
+
+  /// Validates AND, OR special form expression rewrites that can short circuit
+  /// expression evaluation.
+  /// @param input Input SQL expression to be rewritten.
+  /// @param expected Expected SQL expression after rewriting `input`.
+  /// @param type Type of `input` and `expected`.
+  void testRewrite(
+      const std::string& input,
+      const std::string& expected,
+      const RowTypePtr& type = ROW({})) {
+    const auto expr = makeTypedExpr(input, type);
+    const auto rewritten =
+        expression::ExprRewriteRegistry::instance().rewrite(expr);
+    const auto expectedExpr = makeTypedExpr(expected, type);
+    SCOPED_TRACE(fmt::format("Input: {}", expr->toString()));
+    SCOPED_TRACE(fmt::format("Rewritten: {}", rewritten->toString()));
+    SCOPED_TRACE(fmt::format("Expected: {}", expectedExpr->toString()));
+
+    ASSERT_TRUE(*rewritten == *expectedExpr);
+  }
+};
+
+TEST_F(ConjunctRewriteTest, conjunct) {
+  testRewrite("true and true", "true");
+  testRewrite("false or false", "false");
+  testRewrite("null::boolean and false", "false");
+  testRewrite("true or null::boolean", "true");
+  testRewrite(
+      "null::boolean and null::boolean and null::boolean", "null::boolean");
+  testRewrite(
+      "null::boolean or null::boolean or null::boolean", "null::boolean");
+  testRewrite("null::boolean and true", "null::boolean");
+  testRewrite("false or null::boolean", "null::boolean");
+
+  auto type = ROW({"a"}, {VARCHAR()});
+  testRewrite(
+      "null::boolean and a = 'z' and true", "null::boolean and a = 'z'", type);
+  testRewrite(
+      "a = 'z' or null::boolean or false", "a = 'z' or null::boolean", type);
+  testRewrite("true and a = 'z' and true", "a = 'z'", type);
+  testRewrite("false or a = 'z' or false", "a = 'z'", type);
+
+  type = ROW({"a", "b"}, {VARCHAR(), BIGINT()});
```

Comment:
```
auto type = ROW({"a"}, {VARCHAR()});
...
type = ROW({"a", "b"}, {VARCHAR(), BIGINT()});
```

No need to use different values of 'type' for different tests. Just use the latter one for all.